vm/src/vm/vm_ops.rs

```diff
@@ -103,371 +132,260 @@ impl VirtualMachine {
-        self.call_or_unsupported(a, b, identifier!(self, __irshift__), |vm, a, b| {
-            vm.call_or_reflection(
-                a,
-                b,
-                identifier!(self, __rshift__),
-                identifier!(self, __rrshift__),
-                |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, ">>=")),
-            )
-        })
-    }
-
-    pub fn _xor(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_reflection(
-            a,
-            b,
-            identifier!(self, __xor__),
-            identifier!(self, __rxor__),
-            |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "^")),
-        )
-    }
-
-    pub fn _ixor(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_unsupported(a, b, identifier!(self, __ixor__), |vm, a, b| {
-            vm.call_or_reflection(
-                a,
-                b,
-                identifier!(self, __xor__),
-                identifier!(self, __rxor__),
-                |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "^=")),
-            )
-        })
+            &PyNumberBinaryOpSlot::InplaceAdd,
+            &PyNumberBinaryOpSlot::Add,
+            &PyNumberBinaryOpSlot::RightAdd,
+        )?;
+        if !result.is(&self.ctx.not_implemented) {
+            return Ok(result);
+        }
+        if let Ok(seq_a) = PySequence::try_protocol(a, self) {
+            let result = seq_a.inplace_concat(b, self)?;
+            if !result.is(&self.ctx.not_implemented) {
+                return Ok(result);
+            }
+        }
+        Err(self.new_unsupported_binop_error(a, b, "+="))
     }
 
-    pub fn _or(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_reflection(
+    pub fn _mul(&self, a: &PyObject, b: &PyObject) -> PyResult {
+        let result = self.binary_op1(
             a,
             b,
-            identifier!(self, __or__),
-            identifier!(self, __ror__),
-            |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "|")),
-        )
-    }
-
-    pub fn _ior(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_unsupported(a, b, identifier!(self, __ior__), |vm, a, b| {
-            vm.call_or_reflection(
-                a,
-                b,
-                identifier!(self, __or__),
-                identifier!(self, __ror__),
-                |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "|=")),
-            )
-        })
+            &PyNumberBinaryOpSlot::Multiply,
+            &PyNumberBinaryOpSlot::RightMultiply,
+        )?;
+        if !result.is(&self.ctx.not_implemented) {
+            return Ok(result);
+        }
+        if let Ok(seq_a) = PySequence::try_protocol(a, self) {
+            let n =
+                b.try_int(self)?.as_bigint().to_isize().ok_or_else(|| {
+                    self.new_overflow_error("repeated bytes are too long".to_owned())
+                })?;
```

Comment:
Currently float number will also be converted to `isize` and used to repeat seq, which is unexpected.
```
>>>>> b'a' * 3.14
b'aaa'
```
The above would fail in CPython with the following error message.
```
TypeError: can't multiply sequence by non-int of type 'float'
```
@youknowone What is the proper way to restrict non-int numbers here please?

Reply:
instead of `try_int`, `try_index` or `try_index_opt` will be helpful

Reply:
`try_int` has been changed to `try_index` to disable repeating sequences by float number.